File: docs/part2/chapter22.py
# Chapter 22 실패 처리하기


class TestResult:
  def __init__(self):
    self.runCount = 0
    self.failedCount = 0
  def testStarted(self):
    self.runCount = self.runCount + 1
  def testFailed(self):
    self.failedCount = self.failedCount + 1
  def summary(self):
    return "%d run, %d failed" % (self.runCount, self.failedCount)

class TestCase:
  def __init__(self, name):
    self.name = name
  def setUp(self):
    pass
  def run(self):
    result = TestResult()
    result.testStarted()
    self.setUp()
    try:
      method = getattr(self, self.name)
      method()
    except:
      result.testFailed()
    self.tearDown()
    return result
  def tearDown(self):
    pass

class WasRun(TestCase):
  def __init__(self, name):
    self.wasRun = None
    TestCase.__init__(self, name)
  def testMethod(self):
    self.wasRun = 1
    self.log= self.log + "testMethod "
  def setUp(self):
    self.wasRun = None
    self.wasSetUp = 1
    self.log = "setUp "
  def tearDown(self):
    self.log = self.log + "tearDown "
  def testBrokenMethod(self):
    raise Exception

class TestCaseTest (TestCase) :
  def setUp(self) :
    self.test = WasRun("testMethod")
  def testTemplateMethod(self):
    test= WasRun("testMethod")
    test.run()
    assert("setUp testMethod tearDown " == test.log)
  def testResult(self):
    test= WasRun("testMethod")
    result = test.run()
    assert("1 run, 0 failed" == result.summary())
  def testFailedResult(self):
    result = TestResult()
    result.testStarted()
    result.testFailed()
    assert("1 run, 1 failed" == result.summary())

TestCaseTest("testTemplateMethod").run()
TestCaseTest("testResult").run()

File: docs/part2/chapter22.js
class TestResult {
  constructor() {
    this.runCount = 0;
    this.failedCount = 0;
  }

  testStarted = () => {
    this.runCount += 1;
  };

  testFailed = () => {
    this.failedCount += 1;
  };

  summary = () => {
    return `${this.runCount} run, ${this.failedCount} failed`;
  };
}

class TestCase {
  constructor(name) {
    this.name = name;
  }
  setUp = () => {
    return;
  };
  run = () => {
    const result = new TestResult();
    result.testStarted();

    this.setUp();

    try {
      if (!this[this.name]) return;
      const method = this[this.name];
      method();
    } catch {
      result.testFailed();
    }

    this.tearDown();
    return result;
  };

  tearDown = () => {
    return;
  };
}

class WasRun extends TestCase {
  constructor(name) {
    super(name);
    this.wasRun = 0;
  }

  testMethod = () => {
    this.wasRun = 1;
    this.log += "testMethod ";
  };

  setUp = () => {
    this.wasRun = 0;
    this.wasSetUp = 1;
    this.log = "setUp ";
  };

  tearDown = () => {
    this.log += "tearDown";
  };
}

class TestCaseTest extends TestCase {
  constructor(name) {
    super(name);
  }

  setUp = () => {
    this.test = new WasRun("testMethod");
  };
  testTemplateMethod = () => {
    const test = new WasRun("testMethod");
    test.run();
    console.log("setUp testMethod tearDown" === test.log);
  };

  testResult = () => {
    const test = new WasRun("testMethod");
    const result = test.run();
    console.log("1 run, 0 failed" == result.summary());
  };

  testFailedResult = () => {
    const test = new WasRun("testBrokenMethod");
    const result = test.run();
    console.log("1 run, 1 failed" === result.summary());
  };
}

const test1 = new TestCaseTest("testTemplateMethod");
const test2 = new TestCaseTest("testResult");

test1.run();
test2.run();
